### PR TITLE
[WIP] Check checksum of all files of a app when installing it

### PR DESCRIPTION
**Pull request by LEDfan**
Refs  #14616
This currently works when:
- enabling an app via the web interface
- enabling an app via the CLI
- when the app is available on the filesystem
- or when the app is downloaded via the AppStore ~~(not sure if so. I'll upload an app to the appstore which should fail when installing)~~ I checked this. The checksums are checked.
- and when it has never been installed before (so the SQL tables aren't in the DB and in the oc_appconfig table)

cc @Raydiation @LukasReschke

(please also review the error messages)
TODO:
- [ ] documentation for developers
- [ ] documentation for end-users and administrators
- [ ] improve errors
- [ ] log when there is no checksum.json file
- [x] also call the checksum stuff in the updateApp method
- [ ] replace copyr with rename
- [ ] add unit tests

### CODE REVIEW

**Comment by Raydiation** [comment]
Looks good but the error messages should also tell the user what the problem was and how he can fix it

**Comment by karlitschek** [comment]
looks good! :+1:  But @Raydiation is right. The error message should be more use friendly.

**Comment by DeepDiver1975** [comment]
unit tests are missing - see https://github.com/owncloud/core/commit/b533aad21697de29e8d69d7bb99b0309f2114d07#diff-6a3371457528722a734f3c51d9238c13

**Comment by LEDfan** [comment]
@DeepDiver1975 I'm aware of that.

How can I test if this works with the Appstore? I created an app (https://apps.owncloud.com/content/show.php/testy?content=169074) which has a wrong checksum. However the app doesn't show up in the Appstore because it isn't approved. I changed the url in the code, but it still doesn't appear.

**Comment by Raydiation** [comment]
BTW: can you post a grunt config?

**Comment by LEDfan** [comment]
@Raydiation here it is:
````Grunftile.js
module.exports = function(grunt) {
	grunt.initConfig({
		pkg: grunt.file.readJSON('package.json'),
		cachebuster: {
			build: {
				options: {
					format: 'json',
				},
				src: [ '**' ], 
				dest: 'appinfo/checksum.json'
			}
		}
	});
	grunt.loadNpmTasks('grunt-cachebuster');
	grunt.registerTask('default', cachebuster']);
};
````

**Comment by Raydiation** [comment]
We probably want to exclude js/node_modules, tests/ and l10n (since l10n PRs wont update the checksums)

**Comment by LEDfan** [comment]
Yup. I updated the gruntfile above and tested in on the News app.

**Comment by Raydiation** [comment]
Ok, implemented it for the news app. I've used a pre-commit hook to regenerate the checksums before commiting because it takes a while :)

https://github.com/owncloud/news/blob/master/bin/git/hooks/pre-commit

If anyone is interested I've got some more blacklisted files in there https://github.com/owncloud/news/blob/master/js/Gruntfile.js#L224

@DeepDiver1975 can we include this already in 8.1? I'm getting tons of bug reports lately about incorrect app store updates that would be fixed with this PR

Will test the feature ASAP :)

**Comment by Raydiation** [comment]
Ok, I've added the appinfo/checksum.json check the news app repository.

I've created all checksums, disabled the app, changed a file (appinfo/application.php) and reenabled the app again, it shows no warnings. @LEDfan am I doing something wrong?

**Comment by LEDfan** [comment]
Currently it only checks the checksums when the DB records/tables aren't available. i.e. when the app is installing, not when it is being enabled.

**Comment by Raydiation** [comment]
That will fail for the most important usecase, namely updating :)

**Comment by Raydiation** [comment]
Can we get this in on app enable/update?

**Comment by LEDfan** [comment]
@Raydiation should be possible. Maybe add a function call here: https://github.com/LEDfan/core/blob/apps-checksum/lib/private/app/appmanager.php#L144-L147.

**Comment by LEDfan** [comment]
I'm not sure that code is called when installing an app? Somtimes this code is just called to enable an app:
````
$config->setAppValue($app, 'enabled', 'yes');
````

**Comment by Raydiation** [comment]
@LEDfan your code is in the wrong section, move it to the checkAppsIntegrity method: https://github.com/LEDfan/core/blob/apps-checksum/lib/private/installer.php#L288

It should be fine to just do the checks on installation and update and leave it out for enable app since it is a bit harder to do that (checkAppsIntegrity is hard to use outside of the installer class)

**Comment by LEDfan** [comment]
@Raydiation thanks for pointing that out! If I place the code there will it be called both on installation and updating, but not when simply enabling and disabling(so when it already was installed a time ago).

Maybe it's an idea to add an `occ` command for this. This way users can manually check an app, if they have a problem with it?

I'll work on this tomorrow.

**Comment by Raydiation** [comment]
No need for an occ command ;)

**Comment by Raydiation** [comment]
BTW: what still kinda bugs me is where these issues occur? After the extraction? When you copy the directory? 

I think we should get rid of the copy and use the move command (http://php.net/manual/en/function.rename.php) since its guaranteed to be atomic, see:
* https://github.com/LEDfan/core/blob/apps-checksum/lib/private/installer.php#L96
* https://github.com/LEDfan/core/blob/apps-checksum/lib/private/installer.php#L197

That way we can check if all files are in place after extraction and using move guarantee that the exact same files that were checked are actually installed into the new directory. A possible side effect should also be that it's way faster.

**Comment by Raydiation** [comment]
So basically: 
* Replace OC_Helper::copyr with rename
* Move the checksum check into checkAppsIntegrity

**Comment by Raydiation** [comment]
Ah yeah, another thing to consider is the user manually uploading files over ftp, so we also need to perform this check. It seems your code is in the right place but you just need to add it to the updateApp method https://github.com/LEDfan/core/blob/apps-checksum/lib/private/app.php#L1130

Sorry for the many posts, this stuff is a bit confusing ^^

The copyr should still be refactored to rename though

TL;DR:
* Keep your PR as is
* Also call the checksum stuff in the updateApp method
* replace copyr with rename

**Comment by Raydiation** [comment]
I've tried to work on this by myself but I give up. The code is littered all over the place, we have hundred different ways to check if the app is actually updateable/installable, we return nulls, booleans, throw exceptions, etc.

The app installer and updater code needs undergo a complete rewrite.

**Comment by LEDfan** [comment]
> I've tried to work on this by myself but I give up. The code is littered all over the place, we have hundred different ways to check if the app is actually updateable/installable, we return nulls, booleans, throw exceptions, etc.

> The app installer and updater code needs undergo a complete rewrite.

I experience the same problems. IMO there should be just one class `AppManager`  with a minimum four public functions: 
 - enableApp
 - disableApp
 - getAllApps
 - getAllEnabledApps

Then when someone tries to enable/install/download or wathnot the AppMananger::enableApp function is called. This function will:
 - checks if there is a directory on the system with the app
 - if not download the app from the appstore
 - checks the dependencies, checksums, code integrity, if it doesn't use OC_* etc.
At last it creates the tablets, and records in `oc_appconfig`, when they don't already exist.

But this way out of this PR. I'm not even sure if a refactor/rewrite is needed and wanted.

**Comment by LEDfan** [comment]
Woohoo :fireworks:  I manages to test installation from the appstore. The checksums are checked! :fireworks: @Raydiation :)

**Comment by LEDfan** [comment]
I'll continue working on this today.

**Comment by Raydiation** [comment]
@LEDfan something that came to my mind when using it: Maybe its better to use one indentation to be able to add options and other stuff later on, e.g.

```js
{
    "checksums": {
    },
    // etc
}
```

**Comment by LEDfan** [comment]
@Raydiation mm, I'm not sure if this can be done with the Grunt plugin?

**Comment by Raydiation** [comment]
https://github.com/felthy/grunt-cachebuster#optionscomplete

**Comment by LEDfan** [comment]
I moved the CheckSum function to the AppManager, because otherwise you can still enable an app with an invalid checksum. It's now also checking the checksums when updating and enabling.

**Comment by LEDfan** [comment]
@Raydiation thanks for the link, it's updated in the latest commit.

**Comment by LEDfan** [comment]
@DeepDiver1975 how can I write unit (or integration) test for this? Since the function uses native PHP functions. I asked @Raydiation the same questions, he says to wait until app management is non static.

**Comment by MorrisJobke** [comment]
cc @LukasReschke

**Comment by LEDfan** [comment]
Someone willing to help me with the unit tests? :smiley:

**Comment by LukasReschke** [comment]
I do have another approach which also involves code signing etc… - I'd prefer to go with my approach. I will submit a PR later.

**Comment by LEDfan** [comment]
@LukasReschke that's even better :smiley:
